### ISSUE TITLE
PHP7-Support via SWIG

### ISSUE THREAD

**Issue by AlexanderGabriel** [mapserver/mapserver#5557]
Hi together,

i did some work to make PHP-MapScript compile with SWIG on Linux. I think, this is the way, we should implement this to have less work in future.
It is a really early state an very poor testet and theres still a little bit of work to do (for example check exact swig version on 3.0.11) I testet in this environment:
- Debian testing (from last week) x64

It completely builds with CMake. To enable PHP7-Support add -DWITH_PHP7=1.
I thought this is a good option to be safe in Future when new PHP-Versions arrive and SWIG does not support them. Maybe i can merge this together into one CMake-Module to autodetect PHP-Version and SWIG but this is Future.

It does not yet compile on Windows. Not sure if anyone ever testet to compile any SWIG-Extension for PHP7 on Windows and this is a bug in SWIG or if there is still anything missing but i thought, it would be good to make the current state available and invite everybody to test this on different environments and to a little QA :)

My branch is here: https://github.com/AlexanderGabriel/mapserver/tree/PHP7-MapScript
Please note, you have to include mapscript.php as already written here to make Classes and Constants available: http://mapserver.org/development/rfc/ms-rfc-117.html

If you have any Feedback or suggestions, write here so we can discuss.

Kind regards,
Alex

**Comment by sdlime** [comment]
Hi Alex: I thought @jmckenna already did this work. How does your work support or build on Jeff's? --Steve

**Comment by AlexanderGabriel** [comment]
Hi Steve,

i'm not sure but it is possible that i missed something.
I hat a look at his rep, linked in the RFC but he made changes to the existing Mapscript-Extension. 
If i understand the RFC correkt, the goal is not to update the existing code but use the swig-configuration for future and remove the old code someday.
I just read again. He suggested to use phpng as prefix i used php7. I can change this or merge everything together and we do some auto-detection.
I asked him per email some days ago but did not yet receive any answer (it was on wednesday so this is okay).

Do you have a link to his work? Google didn't help :(

Alex

**Comment by AlexanderGabriel** [comment]
HI,

i did a little more work and i'm kind of done for now. At https://github.com/AlexanderGabriel/mapserver/tree/PHPNG-Mapscript there is a branch for adding PHP-Support via SWIG.
There is a bug in SWIG 3.0.12 which prevents building the module on Windows with Thread-Safe PHP but this is fixed in swigs current developement state and i have a workaround (build swig latest dev state or edit the wrapper-file by hand - I would explain this in a README.WIN32 when updating documentation after merging the branch) and can build on windows with thread-safe, non-thread-safe x86 and x64 PHP (latest branch and 7.2) aswell as on debian testing.
I tested and can create Mapfiles, add Layers. etc. but i did not do very intensive testing.

My changes add another option WITH_PHPNG, that will build the extension using SWIG.
Minimal required SWIG version ist 3.0.11 because the support for php7.

Swig still throws an error:
```
C:\dev\work\mapserver\mapscript\v8\..\..\mapserver.h(932) : Warning 314: 'empty' is a PHP keyword, renaming to 'c_empty'
```
I did not change this, because i don't want to mess up anything.

If it is okay, i would open a pull request.

**Comment by sdlime** [comment]
Feel free to open a pull request (if you're still interested). I think PHP support via the old PHP/MapScript code AND Swig is probably in the cards... --Steve

**Comment by geomunion** [comment]
Hi,
nice work @AlexanderGabriel ! 👍  I've successfully compiled your branch PHPNG-Mapscript on current Debian Testing. ... And did some basic tests, too.
Would you open a pull request?

@sdlime, is there still a change to get in Mapserver Release 7.2? Is there any deadline?

Greetings,
Tobias

**Comment by AlexanderGabriel** [comment]
Hi,

thank you, tobias.
i'm still intereset. but there is a problem with swig on windows which prevents building thread-safe modules... this is already fixed (https://github.com/swig/swig/commit/224bb9e0232a219b7cefd23472b592b39e277f46) but not yet released. building swig on windows is no fun. This is why i would like to wait until the next swig release... alternatively, i could add something to the readmes for windows so the non-windows-users can already start using the new module.
how did mapserver project handle things like this in the past?

Greetings,
Alex

**Comment by AlexanderGabriel** [comment]
Hi,

i opened the pull-request: https://github.com/mapserver/mapserver/pull/5611
If this helps all the non-windows-users, it is better to have it in the code than waiting (maybe month) for the next swig release. Maybe i can find some time the next few days to update build-instructions.

Kind regards,
Alex

**Comment by geographika** [comment]
@AlexanderGabriel (and others) - I'm not sure if the SWIG version is a MapServer responsibility or not. 

The GISInternals builds are using SWIG 1.3.39 which breaks Python but works with .NET.
SWIG 3.0.12 works with Python but breaks .NET. 
I'm not sure which versions Java and Ruby work with. 

Maybe simply adding a minimum/maximum SWIG version for each MapScript language to the docs is an option?

**Comment by geographika** [comment]
Implemented with #5675
